File: gachisa-backend/src/main/java/com/gachisa/product/entity/Product.java
package com.gachisa.product.entity;

import com.gachisa.category.entity.Category;
import com.gachisa.global.exception.CustomException;
import com.gachisa.global.exception.ErrorCode;
import com.gachisa.user.entity.User;
import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

@Entity
@Table(name = "product")
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Product {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "seller_id", nullable = false)
    private User seller;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id", nullable = false)
    private Category category;

    @Column(nullable = false)
    private String name;

    @Lob
    private String description;

    @Column(nullable = false)
    private int basePrice;

    @Column(nullable = false)
    private int stock;

    private String imageUrl;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private ProductStatus status;

    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Builder
    private Product(User seller, Category category, String name, String description, int basePrice, int stock,
                    String imageUrl, ProductStatus status, LocalDateTime createdAt) {
        this.seller = seller;
        this.category = category;
        this.name = name;
        this.description = description;
        this.basePrice = basePrice;
        this.stock = stock;
        this.imageUrl = imageUrl;
        this.status = status;
        this.createdAt = createdAt;
    }

    public void stopSale() {
        this.status = ProductStatus.SUSPENDED;
    }

    public void resumeSale() {
        this.status = ProductStatus.ON_SALE;
    }

    public boolean isOnSale() {
        return status == ProductStatus.ON_SALE;
    }

    public boolean isOwnedBy(Long userId) {
        return this.seller.getId().equals(userId);
    }

    public void updateName(String name) {
        this.name = name;
    }

    public void updateDescription(String description) {
        this.description = description;
    }

    public void updateBasePrice(int basePrice) {
        this.basePrice = basePrice;
    }

    public void updateStock(int stock) {
        this.stock = stock;
    }

    public void increaseStock(int quantity) {
        this.stock += quantity;
    }

    public void decreaseStock(int quantity) {
        if (this.stock < quantity) {
            throw new CustomException(ErrorCode.INSUFFICIENT_STOCK);
        }
        this.stock -= quantity;
    }

    public void updateCategory(Category category) {
        this.category = category;
    }

    public void updateImageUrl(String imageUrl) {
        this.imageUrl = imageUrl;
    }
}

File: gachisa-backend/src/main/kotlin/com/gachisa/product/entity/Product.kt
package com.gachisa.product.entity

import com.gachisa.category.entity.Category
import com.gachisa.user.entity.User
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.FetchType
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.Lob
import jakarta.persistence.ManyToOne
import jakarta.persistence.Table
import java.time.LocalDateTime

@Entity
@Table(name = "product")
class Product private constructor(
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "seller_id", nullable = false)
    val seller: User,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id", nullable = false)
    var category: Category,

    @Column(nullable = false)
    var name: String,

    @Lob
    var description: String?,

    @Column(nullable = false)
    var basePrice: Int,

    @Column(nullable = false)
    var stock: Int,

    var imageUrl: String?,

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    var status: ProductStatus,

    @Column(nullable = false, updatable = false)
    val createdAt: LocalDateTime,
) {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long? = null
        protected set

    fun stopSale() {
        this.status = ProductStatus.SUSPENDED
    }

    fun resumeSale() {
        this.status = ProductStatus.ON_SALE
    }

    fun isOwnedBy(userId: Long): Boolean = this.seller.id == userId

    fun updateName(name: String) {
        this.name = name
    }

    fun updateDescription(description: String?) {
        this.description = description
    }

    fun updateBasePrice(basePrice: Int) {
        this.basePrice = basePrice
    }

    fun updateStock(stock: Int) {
        this.stock = stock
    }

    fun updateCategory(category: Category) {
        this.category = category
    }

    fun updateImageUrl(imageUrl: String?) {
        this.imageUrl = imageUrl
    }

    companion object {
        @JvmStatic
        fun of(
            seller: User,
            category: Category,
            name: String,
            description: String?,
            basePrice: Int,
            stock: Int,
            imageUrl: String?,
            status: ProductStatus,
            createdAt: LocalDateTime,
        ): Product = Product(seller, category, name, description, basePrice, stock, imageUrl, status, createdAt)
    }
}
